Validate that the prototype is compatible with the new content node

carlobeltrame opened the pull request:
Prevents copying a SafetyConcept node from a ColumnLayout node (backend strategy class must match), and also prevents copying a SafetyConcept node from a Notes node (content type must match).

carlobeltrame committed:
Validate that the prototype is compatible with the new content node

usu commented on api/src/Validator/ContentNode/AssertPrototypeCompatibleValidator.php:
```diff
@@ -0,0 +1,51 @@
+
+        $class = get_class($object);
+
+        if (!is_a($value, $class, false)) {
```
Wollen wir hier nicht auf Gleichheit überprüfen? Macht im Moment zwar noch kein Unterschied. Aber nur für den Fall, dass wir irgendwann auch mal bestehende ContentNodes extenden.

carlobeltrame replied:
Das habe ich mir auch überlegt, aber ich dachte Subklasse müsste auch erlaubt sein..? Für ein ColumnLayout müsste es doch erlaubt sein, einen Prototype vom Typ ThreeColumnLayout zu verwenden, nicht? Der Check im Datapersister würde das ja auch akzeptieren.

usu replied:
Hmm, ja, stimmt. Ich hatte zuerst gedacht es wär falsch rum, aber stimmt eigentlich so. Konzeptuell könnte ich ein neues ColumnLayout erstellen und dabei die Daten von einem prototype ThreeColumnLayout kopieren.

Nicht sicher, ob wir wirklich mal so einen Use-case haben werden. Ich hätte jetzt eher dazu tendiert, hier stricter zu sein (prüfen auf Gleichheit) und implementierten, wenn wir den Use-case wirklich haben. Kann aber mit beidem leben.

Wenn wir's so wie jetzt lassen, braucht es auf jeden Fall die zusätzliche Überprüfung, dass der ContentType kompatibel ist. Sonst kann ich dem ColumnLayout einen ContentType unterjubeln, der nur auf ThreeColumnLayout passt.

usu commented on api/src/Validator/ContentNode/AssertPrototypeCompatibleValidator.php:
```diff
@@ -0,0 +1,51 @@
+        $actualContentType = $value->getContentTypeName();
+        $expectedContentType = $this->context->getObject()->getContentTypeName();
+
+        if ($expectedContentType !== $actualContentType) {
```
`expectedContentType` könnte auch `null` sein. In diesem Fall wird der contentType vom prototype kopiert.

Dann müsste man nur prüfen, dass `$prototype->contentType->entityClass` kompatibel ist mit `$object::class`. Das ist aber eigentlich durch den Test oben bereits abgedeckt.

- `$prototype->contentType->entityClass` ist kompatibel mit `$prototype::class` (kann gar nicht anders in der DB stehen)
- $prototype::class ist gleich wie $object::class (wird oben geprüft)

carlobeltrame replied:
Stimmt. Ich könnte höchstens die AssertContentTypeCompatible Validierung auf einen neuen Getter getContentTypeFromThisOrPrototype verschieben. Dann wäre sogar inkonsistenter Datenbankzustand abgedeckt, man könnte dann den inkonsistenten ContentNode nicht mehr als Prototype einsetzen. Fändest du das sinnvoll?

usu replied:
Finde ich einen guten Vorschlag. Und deckt den Edge-case oben auch ab.

carlobeltrame replied:
Ich habe das so ergänzt. Allerdings ist es im Moment doch noch nicht möglich, dass `expectedContentType` `null` ist, weil die Symfony AutoMapping Validierung auf `contentType` noch greift (weil die DB-Spalte non-nullable ist). Ich habe ein `DisableAutoMapping` auf `ContentNode::$contentType` hinzugefügt, aber wegen einem [Bug in Symfony](https://github.com/symfony/symfony/issues/35399) reicht das nicht, wir müssten das auf allen Subklassen auch nochmals schreiben. Stattdessen habe ich jetzt geskippte Tests dafür vorbereitet, die wir dann aktivieren können wenn Symfony den Bug gefixt hat. Ein [PR](https://github.com/symfony/symfony/pull/40294) dafür ist seit Februar offen, aber es ist anscheinend nicht ganz einfach, das zu fixen.

carlobeltrame committed:
Handle edge case when no contentType is sent

This specific case is not supported yet because of symfony bug
https://github.com/symfony/symfony/issues/35399, but we have prepared
tests for when this bug is resolved.

Also moves some more ContentNode tests to the test superclass.

Also slightly changes a validation message to be easier to understand.
